### PR TITLE
Fix #3773: Fix for junit and roboelectric tests not running in android studio artic fox

### LINKED ISSUE #3773: roboelectric and Junit tests not working in latest android studio arctic fox
**Describe the bug**
As you go to run any specific test by adding configuration a weird error comes saying that the memory is Filled fully 
this is the SS of the error
image
image (1)

**To Reproduce**
image
click on it then on edit configuration

After that a window pops up like this click on + symbol at the top left corner and then click on junit
image

then this window appears
image

then click on the blue part and select a app module and search a class  to retrace what I did select the app module and then search the class StoryFragmentTest select it and click on apply and Ok then run the test by clicking on the run button or by shortcut Shift + F10 and you're done wait for it then the error appears as the gradle compiles and the app runs
NOTE that in android studio 4.2.1 this successfully runs this error happens only in artic fox

**Expected behavior**

**Screenshots**
If applicable, add screenshots or videos to help explain your problem.

**Device**
- Device/emulator being used
- SDK version

**Additional context**

### PR DESCRIPTION
**Pull request by yash10019coder**
## Explanation

Fixes #3773 
Added a line in gradle.property due to which heap size is increased and also increased the heap size for unitTest in gradle build app

Before | After 
----------|-------
   |  test_passed
# Error code for the old code when running the tests
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at org.jetbrains.kotlin.com.intellij.openapi.util.io.FileUtilRt.loadBytes(FileUtilRt.java:847)
	at org.jetbrains.kotlin.com.intellij.openapi.util.io.FileUtil.loadBytes(FileUtil.java:1409)
	at org.jetbrains.kotlin.com.intellij.openapi.vfs.impl.ZipHandlerBase.contentsToByteArray(ZipHandlerBase.java:146)
	at org.jetbrains.kotlin.com.intellij.openapi.vfs.impl.jar.CoreJarVirtualFile.contentsToByteArray(CoreJarVirtualFile.java:127)
	at org.jetbrains.kotlin.com.intellij.openapi.vfs.VirtualFile.contentsToByteArray(VirtualFile.java:606)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClsFileImpl$1.findInnerClass(ClsFileImpl.java:622)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClsFileImpl$1.findInnerClass(ClsFileImpl.java:612)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.StubBuildingVisitor.visitInnerClass(StubBuildingVisitor.java:270)
	at org.jetbrains.org.objectweb.asm.ClassReader.accept(ClassReader.java:662)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClsFileImpl.buildFileStub(ClsFileImpl.java:574)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClassFileStubBuilder.lambda$buildStubTree$1(ClassFileStubBuilder.java:69)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClassFileStubBuilder$$Lambda$269/1237933695.get(Unknown Source)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClassFileStubBuilder.setContentAndCompute(ClassFileStubBuilder.java:105)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClassFileStubBuilder.buildStubTree(ClassFileStubBuilder.java:57)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClassFileStubBuilder.buildStubTree(ClassFileStubBuilder.java:24)
	at org.jetbrains.kotlin.com.intellij.psi.stubs.BinaryFileStubBuilder$CompositeBinaryFileStubBuilder.buildStubTree(BinaryFileStubBuilder.java:53)
	at org.jetbrains.kotlin.com.intellij.psi.stubs.StubTreeBuilder.buildStubTree(StubTreeBuilder.java:105)
	at org.jetbrains.kotlin.com.intellij.psi.stubs.StubTreeBuilder.buildStubTree(StubTreeBuilder.java:90)
	at org.jetbrains.kotlin.com.intellij.psi.stubs.CoreStubTreeLoader.readOrBuild(CoreStubTreeLoader.java:48)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClsFileImpl.getStubTree(ClsFileImpl.java:480)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClsFileImpl.getStub(ClsFileImpl.java:452)
	at org.jetbrains.kotlin.com.intellij.psi.impl.compiled.ClsFileImpl.getClasses(ClsFileImpl.java:133)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl$Companion.findClassInPsiFile(KotlinCliJavaFileManagerImpl.kt:261)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl$Companion.access$findClassInPsiFile(KotlinCliJavaFileManagerImpl.kt:257)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl.findPsiClassInVirtualFile(KotlinCliJavaFileManagerImpl.kt:227)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl.access$findPsiClassInVirtualFile(KotlinCliJavaFileManagerImpl.kt:48)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl$findPsiClass$1.invoke(KotlinCliJavaFileManagerImpl.kt:70)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl$findPsiClass$1.invoke(KotlinCliJavaFileManagerImpl.kt:48)
	at org.jetbrains.kotlin.util.PerformanceCounter.time(PerformanceCounter.kt:101)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl.findPsiClass(KotlinCliJavaFileManagerImpl.kt:69)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCliJavaFileManagerImpl.findClass(KotlinCliJavaFileManagerImpl.kt:135)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiElementFinderImpl.findClass(PsiElementFinderImpl.java:44)

         

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by rt4914** [comment]
@yash10019coder Please add before/after screenshot of test on robolectirc and espresso here. 
Note: The before screenshots should contain error msg so that Ben can check it.

File: app/build.gradle [hunk 1]
```diff
@@ -56,7 +56,7 @@ android {
     unitTests {
       includeAndroidResources = true
       all {
-        maxHeapSize = "1024m"
+        maxHeapSize = "4024m"
```
Comment (BenHenning) [inline app/build.gradle]: Probably should use 4096 (it's generally conventional to use multiples of '1024').

File: gradle.properties [hunk 2]
```diff
@@ -9,8 +9,8 @@
 
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
-org.gradle.jvmargs=-Xmx1024m
-
+#org.gradle.jvmargs=-Xmx1024m
+org.gradle.jvmargs=-Xmx4096m -XX:MaxPermSize=4096m -XX:+HeapDumpOnOutOfMemoryError
```
Comment (BenHenning) [inline gradle.properties]: Why enable HeapDumpOnOutOfMemoryError?

Also, do we need both MaxPermSize & Xmx?

File: gradle.properties [hunk 3]
```diff
@@ -9,8 +9,8 @@
 
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
-org.gradle.jvmargs=-Xmx1024m
-
+#org.gradle.jvmargs=-Xmx1024m
```
Comment (BenHenning) [inline gradle.properties]: Prefer not to comment out code--remove/update it instead.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Sorry for the delay here.

Thanks @yash10019coder for filing & investigating this! I've hit the GC limit when running from Gradle CLI (I mainly use Bazel for development within AS), with Arctic Fox running. Do we know why this is tied to Arctic Fox? That doesn't make much sense to me since Gradle runs outside it. Maybe it's just Arctic Fox is consuming more resources & we're already close to breaking the 1024MB limit?

Either way, the change seems reasonable. My main concern is this will have a higher chance of locking people's systems if they're running with low ram, but 4G is a reasonable ask for an Android development environment, I think.

Had a couple of other comments.

**Comment by rt4914** [comment]
@yash10019coder and @BenHenning FYI this solution did not work for @darkmat13r and he was facing same errors.

**Comment by yash10019coder** [comment]
> @yash10019coder and @BenHenning FYI this solution did not work for @darkmat13r and he was facing same errors.

@rt4914 @darkmat13r can you please provide the error log

**Comment by BenHenning** [comment]
@rt4914 or @darkmat13r can you please follow up with https://github.com/oppia/oppia-android/pull/3783#issuecomment-927002504? I'm kind of keen to get this in because it's affected me as well. Gradle builds are really non-performant in AS Arctic Fox at the moment (to the point where they can actually stall indefinitely for larger build/test runs).

**Comment by darkmat13r** [comment]
@yash10019coder   Now getting this Error

image

**Comment by BenHenning** [comment]
@yash10019coder is this something that you're still working on? It would be really nice to see this get wrapped up since it's affecting contributors.

If you don't have the bandwidth to take this, please let us know (reassigning the base issue is no problem, we just want to make sure that folks assigned are actively working on their issues).

**Comment by yash10019coder** [comment]
Hi, @BenHenning apologies for not responding to this PR I got busy with other PR's and I totally forgot about this one.
I think that this issue is fixed We are getting this error while running tests with JUnit and roboelectric on the latest arctic fox as well as the earlier IntelliJ version of the android studio.
This is the gist link for the error we are getting [link](https://gist.github.com/yash10019coder/810609367a319b1cdccc38f4fb5a2325)
and this error exists while running any test case and on all of the versions of the android studio
Thanks

File: gradle.properties (hunk 3 above)
Comment (yash10019coder) [reply gradle.properties]: removed the commented code

File: gradle.properties (hunk 2 above)
Comment (yash10019coder) [reply gradle.properties]: I think we should use it as if heap runs out of memory it gives us the error and also dumps the memory out

**Review by yash10019coder — COMMENTED** [review]
Hi @BenHenning PTAL

File: gradle.properties (hunk 2 above)
Comment (BenHenning) [reply gradle.properties]: What would we use it for? It's probably a really expensive operation, and the memory dump presumably uses a large amount of disk space. We probably shouldn't enable it if it isn't needed.

I'm actually wondering if all we need here is the 'Xmx' setting.

**Review by BenHenning — COMMENTED** [review]
Thanks for reviving this @yash10019coder. Left one follow-up.

**Comment by BenHenning** [comment]
FYI I'm listed as a codeowner for this PR & I'll be unavailable to perform code reviews over the next 2 weeks--thanks for your patience.

**Comment by darkmat13r** [comment]
@yash10019coder  @BenHenning @rt4914 
I am  still not able to run the test cases. Please check the error in attached image.
image

**Comment by yash10019coder** [comment]
Hi @darkmat13r, I think that while initializing the test you might not have added the correct module
image
image

**Comment by darkmat13r** [comment]
@yash10019coder thank you. let me check that again.

**Comment by darkmat13r** [comment]
@yash10019coder 

Error for configuration are as follows
1.)
image
image

2.)
image
image

**Comment by yash10019coder** [comment]
You shall hide the shorten command line option
By clicking on Modify options
then try again

**Comment by darkmat13r** [comment]
> You shall hide the shorten command line option By clicking on Modify options then try again

Yes, That's the first configuration. By default when you hide Shorten command line option the value is use-local default none. 
And to confirm I tried hiding the the Shorten command line it is giving me same error as first error that I shared.
Please can you share you android studio version.
Mine is 
image

**Comment by yash10019coder** [comment]
image

**Comment by darkmat13r** [comment]
> image

Let me rollback to this version and check.

**Comment by darkmat13r** [comment]
@yash10019coder Which OS are you using?

**Comment by yash10019coder** [comment]
> @yash10019coder Which OS are you using?

image

**Comment by darkmat13r** [comment]
@yash10019coder Same issue with patch 2. I will create a Hyper VM for ubuntu 21 and try there.

**Comment by yash10019coder** [comment]
ok



[image: Mailtrack]
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality11&>
Sender
notified by
Mailtrack
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality11&>
11/25/21,
10:37:57 AM

**Comment by BenHenning** [comment]
Hi. FYI I've been out the last couple of weeks, so I'm working to catch up on my reviews. I might be delayed a couple of days, but I'll be reviewing this soon. Thanks for your patience!

**Comment by BenHenning** [comment]
@yash10019coder apologies for the delay on reviewing this.

I noticed that my latest discussion comment wasn't replied to--can you make sure all open threads are replied to before sending this back for a follow-up review? It's hard for me to know whether you decided not to change per my suggestion & why that's the correct choice without an explanation.

File: app/build.gradle (hunk 1 above)
Comment (yash10019coder) [reply app/build.gradle]: done

File: gradle.properties (hunk 3 above)
Comment (yash10019coder) [reply gradle.properties]: done

File: gradle.properties (hunk 2 above)
Comment (yash10019coder) [reply gradle.properties]: ok, I'm using then only -Xmx

**Comment by yash10019coder** [comment]
> @yash10019coder apologies for the delay on reviewing this.
> 
> I noticed that my latest discussion comment wasn't replied to--can you make sure all open threads are replied to before sending this back for a follow-up review? It's hard for me to know whether you decided not to change per my suggestion & why that's the correct choice without an explanation.

sorry for this I forgot the above threads from now onward I will reply to every comment.

**Review by yash10019coder — COMMENTED** [review]
Hi @BenHenning mentioned all the comments all the tests are working and pasing for all of us only for @darkmat13r this isn't happening everthing else is clean PTAL.
Thanks

**Comment by yash10019coder** [comment]
Hi @darkmat13r @BenHenning I ran the tests for espresso junit in my windows 10
image
you can see the version of my android studio and also all of the tests are passing and running correctly
the thing I found is that use this option of shorten command line option by this tests will run in all other configurations this will fail.
Thanks
image

**Review by BenHenning — APPROVED** [review]
Thanks @yash10019coder. LGTM!

**Comment by BenHenning** [comment]
> Hi @darkmat13r @BenHenning I ran the tests for espresso junit in my windows 10 image you can see the version of my android studio and also all of the tests are passing and running correctly the thing I found is that use this option of shorten command line option by this tests will run in all other configurations this will fail. Thanks image

@yash10019coder Android Studio changed how it manages JUnit tests in this version, and there are now some compatibility issues. We may need to document these differences (I was hoping that Bazel would be farther along since then we could switch to it since the Android Studio with Bazel plugin is unaffected by this change, but we'll be in this in-between state for a while longer yet). /cc @FareesHussain as FYI since it's likely we'll see more people confused by how to run JUnit tests in Arctic Fox+ as more people install the latest version of Android Studio.

**Comment by FareesHussain** [comment]
Is it just oppia or all projects?

**Comment by FareesHussain** [comment]
What about using command line?

**Comment by BenHenning** [comment]
I think it has to do with the fact that we define shared Robolectric/Espresso tests, so running JUnit tests directly from the editor just fails. CLI would work, but it makes debugging a lot harder so I think instead we might want to consider updating the testing instructions to have a step-by-step guide for how to add/run JUnit tests in Arctic Fox (this needs some investigation--I forget what the exact issues are since I rarely use Gradle in development).

**Comment by darkmat13r** [comment]
@yash10019coder @BenHenning I think the issue i am getting must be due to fact that I installed android studio from jetbrains toolbox which installs the android in the folder of its data folder so it create a long path for command line that throws error (Command line is too long) in default configuration for the tests. I will install the android studio from installer and see if thats the case.
image

**Comment by BenHenning** [comment]
Ah interesting @darkmat13r. Is that a usual way to install Android Studio? I thought most people just installed it from the downloadable installer from the Android Studio download page.

Also, is it an option for you to reinstall AS using the installer? It sounds like that might fix the problem that you're seeing.

**Comment by darkmat13r** [comment]
@BenHenning No, I just prefer jetbrains toolbox  because its easier to update the IDEs and I use other IDE from jetbrains also. 
Yes I installed it from downloadable installer. The command line error was gone. Now I am getting another error I will share it today. Don't know why that's happening but I will figure that out.

**Comment by darkmat13r** [comment]
@yash10019coder    @BenHenning I am getting this error after running from downloadable android studio

image

**Comment by yash10019coder** [comment]
> @yash10019coder    @BenHenning I am getting this error after running from downloadable android studio
> 
> image
> 

@darkmat13r are using the shorten command line option i mentioned above?

**Comment by darkmat13r** [comment]
> > @yash10019coder    @BenHenning I am getting this error after running from downloadable android studio
> > image
> 
> @darkmat13r are using the shorten command line option i mentioned above?

Yes, Same configuration as yours

**Pull request merged**